### PR TITLE
Improve Dynatrace Exporter logging

### PR DESCRIPTION
**Pull request by evan-bradley**
**Description:**

Improve Dynatrace Exporter logging.

**Testing:**

Unit tests.

### CODE REVIEW

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 1]
```diff
@@ -275,6 +275,16 @@ func (e *exporter) sendBatch(ctx context.Context, lines []string) error {
...
 	}
 
+	if resp.StatusCode == http.StatusTooManyRequests {
+		return consumererror.NewPermanent(
+			fmt.Errorf("The server responded that too many requests have been sent. Please check your export interval and batch sizes, or see the following for more information: https://www.dynatrace.com/support/help/dynatrace-api/basics/access-limit"),
+		)
+	}
+
+	if resp.StatusCode >= http.StatusBadRequest {
+		return consumererror.NewPermanent(fmt.Errorf(`Received error response status: "%v"`, resp.Status))
```
Comment (pirgeo) [inline exporter/dynatraceexporter/metrics_exporter.go:285]: This just prints the status code right? Maybe: 
```suggestion
		return consumererror.NewPermanent(fmt.Errorf(`Received error response status code: "%v"`, resp.Status))
Write
```

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 2]
```diff
@@ -275,6 +275,16 @@ func (e *exporter) sendBatch(ctx context.Context, lines []string) error {
 		return consumererror.NewPermanent(fmt.Errorf("metrics ingest v2 module not found - ensure module is enabled and endpoint is correct"))
 	}
 
+	if resp.StatusCode == http.StatusTooManyRequests {
+		return consumererror.NewPermanent(
+			fmt.Errorf("The server responded that too many requests have been sent. Please check your export interval and batch sizes, or see the following for more information: https://www.dynatrace.com/support/help/dynatrace-api/basics/access-limit"),
```
Comment (pirgeo) [inline exporter/dynatraceexporter/metrics_exporter.go:280]:
```suggestion
			fmt.Errorf("The server responded that too many requests have been sent. Please check your export interval and batch sizes and see https://www.dynatrace.com/support/help/dynatrace-api/basics/access-limit for more information"),
Write
```

**Review by pirgeo — APPROVED** [review]
Just some text nits, feel free to ignore them. I feel like it reads easier that way.

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 1 above)
Comment (evan-bradley) [reply exporter/dynatraceexporter/metrics_exporter.go:285]: This actually prints the status code and status message, e.g. "200 OK". You can see an example in [this assertion](https://github.com/dynatrace-oss-contrib/opentelemetry-collector-contrib/blob/40e819347efddb4f52536ce665d6c1774248d3e8/exporter/dynatraceexporter/metrics_exporter_test.go#L582).

Comment (pirgeo) [reply exporter/dynatraceexporter/metrics_exporter.go:275]: Ah perfect, ignore my comment then!

**Review by pirgeo — APPROVED** [review]
Looking good, thanks for updating the success scenario as well. Everything looks good, but maybe the PR needs a title update.

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 3]
```diff
@@ -275,6 +265,24 @@ func (e *exporter) sendBatch(ctx context.Context, lines []string) error {
...
 	}
 
+	if resp.StatusCode == http.StatusTooManyRequests {
+		return consumererror.NewPermanent(
+			fmt.Errorf("The server responded that too many requests have been sent. Please check your export interval and batch sizes and see https://www.dynatrace.com/support/help/dynatrace-api/basics/access-limit for more information"),
+		)
+	}
+
+	if resp.StatusCode >= http.StatusBadRequest {
+		return consumererror.NewPermanent(fmt.Errorf(`Received error response status: "%v"`, resp.Status))
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:275]: Even if only _some_ data points are invalid (e.g. metric key too long or only invalid characters), `400 Bad Request` will be returned, while the valid data points are still accepted:
https://www.dynatrace.com/support/help/dynatrace-api/environment-api/metric-v2/post-ingest-metrics#response-codes
Would the permanent error only prevent those from being retried or would it disable the entire exporter?

Comment (evan-bradley) [reply exporter/dynatraceexporter/metrics_exporter.go:275]: A `400` status code returned by Dynatrace receives special handling and wont hit this case, I simply started at 400 to be entirely inclusive of all 4xx+ errors. I can update the starting error code or add a comment if this is unclear.

A permanent error is intended to indicate that the same input will always result in the same output, so the Collector shouldn't try to process the same telemetry data again. A permanent error doesn't disable the exporter, we have to set the `isDisabled` property to disable it, and we have unit tests that ensure we only set it when we intend to.

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 4]
```diff
@@ -275,6 +265,24 @@ func (e *exporter) sendBatch(ctx context.Context, lines []string) error {
...
+	}
+
+	if rbUnmarshalErr == nil {
+		e.settings.Logger.Debug(
+			"Response from Dynatrace",
```
[the remark this answers is not in the document]
Comment (evan-bradley) [reply exporter/dynatraceexporter/metrics_exporter.go]: Sounds good to me.

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 3 above)
Comment (arminru) [reply exporter/dynatraceexporter/metrics_exporter.go:275]: Ah sorry, I only saw that clause above now. I think making it a `>` would make it more obvious (although a comment "400 handled above" in addition wouldn't hurt).

Comment (arminru) [reply exporter/dynatraceexporter/metrics_exporter.go:275]:
```suggestion
	if resp.StatusCode > http.StatusBadRequest { // "400 Bad Request" itself is handled above
		return consumererror.NewPermanent(fmt.Errorf(`Received error response status: "%v"`, resp.Status))
```

**Review by arminru — APPROVED** [review]

**Review by pichlermarc — APPROVED** [review]

**Pull request merged**